src/Npgsql/NpgsqlReadBuffer.cs

```diff
@@ -134,17 +182,33 @@ async Task EnsureLong()
                         totalRead += read;
                     }
 
+                    // Resetting cancellation token source, so we can use it again
+                    if (async)
+                        _timeoutCts.CancelAfter(-1);
+
                     NpgsqlEventSource.Log.BytesRead(totalRead);
                 }
                 // We have a special case when reading async notifications - a timeout may be normal
                 // shouldn't be fatal
                 // Note that mono throws SocketException with the wrong error (see #1330)
+                // sync timeout
                 catch (IOException e) when (
-                    dontBreakOnTimeouts && (e.InnerException as SocketException)?.SocketErrorCode ==
+                    (e.InnerException as SocketException)?.SocketErrorCode ==
                        (Type.GetType("Mono.Runtime") == null ? SocketError.TimedOut : SocketError.WouldBlock)
                 )
                 {
-                    throw new TimeoutException("Timeout while reading from stream");
+                    if (dontBreakOnTimeouts)
```

Comment:
This changes the behavior on timeout to not wrap TimeoutException by NpgsqlException, any reason to do this? I'd rather this PR only dealt with frontend async methods, we can leave other unrelated changes to later PRs (you can open an issue for discussion if you think something else can be improved/fixed).

Reply:
I see your point - I'll create another issue some time later.